src/shogun/classifier/Perceptron.cpp

```diff
@@ -78,44 +58,34 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	auto pb = SG_PROGRESS(range(max_iter));
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+	m_output = SGVector<float64_t>(m_continue_features->get_num_vectors());
+}
+
+void CPerceptron::iteration()
+{
+	bool converged = true;
+	SGVector<float64_t> w = get_w();
+
+	auto labels = binary_labels(m_labels)->get_int_labels();
+	auto iter_train_labels = labels.begin();
+	auto iter_output = m_output.begin();
+
+	for (const auto& v : DotIterator(m_continue_features->as<CDotFeatures>()))
 	{
-		COMPUTATION_CONTROLLERS
-		converged=true;
-		auto iter_train_labels = train_labels.begin();
-		auto iter_output = output.begin();
-		for (const auto& v : DotIterator(features))
+		const auto true_label = *(iter_train_labels++);
+		auto& predicted_label = *(iter_output++);
+
+		predicted_label = v.dot(w) + bias;
+
+		if (CMath::sign<float64_t>(predicted_label) != true_label)
```

Comment:
the loop here is such that i think we will need a local variable to check when m_complete needs to be changed( that is check if we never enter the if statement here) if that is so i think the converged bool works fine or?

Reply:
i dont understand, see below